ports/nrf/common-hal/alarm/__init__.c

```diff
@@ -0,0 +1,236 @@
+    }
+    if (reset_reason_saved & NRF_POWER_RESETREAS_RESETPIN_MASK) {
+        return NRF_SLEEP_WAKEUP_RESETPIN;
+    }
+    else if (reset_reason_saved & NRF_POWER_RESETREAS_OFF_MASK) {
+        return NRF_SLEEP_WAKEUP_GPIO;
+    }
+    else if (reset_reason_saved & NRF_POWER_RESETREAS_VBUS_MASK) {
+        return NRF_SLEEP_WAKEUP_VBUS;
+    }
+    return NRF_SLEEP_WAKEUP_UNDEFINED;
+}
+
+#ifdef NRF_DEBUG_PRINT
+static const char* cause_str[] = {
+  "UNDEFINED",
+  "GPIO",
+  "TIMER",
+  "TOUCHPAD",
+  "VBUS",
+  "RESETPIN",
+};
+void print_wakeup_cause(nrf_sleep_source_t cause) {
+    if (cause >= 0 && cause < NRF_SLEEP_WAKEUP_ZZZ) {
+        dbg_printf("wakeup cause = NRF_SLEEP_WAKEUP_%s\r\n",
+		   cause_str[(int)cause]);
+    }
+}
+#endif
+
+bool alarm_woken_from_sleep(void) {
+   nrf_sleep_source_t cause = _get_wakeup_cause();
+#ifdef NRF_DEBUG_PRINT
+   if (cause != NRF_SLEEP_WAKEUP_UNDEFINED) {
+       //print_wakeup_cause(cause);
+   }
+#endif
+   return (cause == NRF_SLEEP_WAKEUP_GPIO || cause == NRF_SLEEP_WAKEUP_TIMER
+	   || cause == NRF_SLEEP_WAKEUP_TOUCHPAD);
+}
+
+STATIC mp_obj_t _get_wake_alarm(size_t n_alarms, const mp_obj_t *alarms) {
+    nrf_sleep_source_t cause = _get_wakeup_cause();
+#ifdef NRF_DEBUG_PRINT
+    //print_wakeup_cause(cause);
+#endif
+    switch (cause) {
+      case NRF_SLEEP_WAKEUP_TIMER: {
+        return alarm_time_timealarm_get_wakeup_alarm(n_alarms, alarms);
+      }
+      case NRF_SLEEP_WAKEUP_TOUCHPAD: {
+        return alarm_touch_touchalarm_get_wakeup_alarm(n_alarms, alarms);
+      }
+      case NRF_SLEEP_WAKEUP_GPIO: {
+        return alarm_pin_pinalarm_get_wakeup_alarm(n_alarms, alarms);
+      }
+      default:
+        break;
+    }
+    return mp_const_none;
+}
+
+mp_obj_t common_hal_alarm_get_wake_alarm(void) {
+    mp_obj_t obj = _get_wake_alarm(0, NULL);
+    return obj;
+}
+
+// Set up light sleep or deep sleep alarms.
+STATIC void _setup_sleep_alarms(bool deep_sleep, size_t n_alarms, const mp_obj_t *alarms) {
+    alarm_pin_pinalarm_set_alarms(deep_sleep, n_alarms, alarms);
+    alarm_time_timealarm_set_alarms(deep_sleep, n_alarms, alarms);
+    alarm_touch_touchalarm_set_alarm(deep_sleep, n_alarms, alarms);
+}
+
+STATIC void _idle_until_alarm(void) {
+#ifdef NRF_DEBUG_PRINT
+    int ct = 40;
+#endif
+    reset_reason_saved = 0;
+    // Poll for alarms.
+    while (!mp_hal_is_interrupted()) {
+        RUN_BACKGROUND_TASKS;
+        // Allow ctrl-C interrupt.
+        if (alarm_woken_from_sleep()) {
+	    alarm_save_wake_alarm();
+	    return;
+	}
+        port_idle_until_interrupt();
+#ifdef NRF_DEBUG_PRINT
+        if (ct > 0) {
+            dbg_printf("_");
+            --ct;
+        }
+#endif
+    }
+}
+
+mp_obj_t common_hal_alarm_light_sleep_until_alarms(size_t n_alarms, const mp_obj_t *alarms) {
+    mp_obj_t r_obj = mp_const_none;
+    _setup_sleep_alarms(false, n_alarms, alarms);
+#ifdef NRF_DEBUG_PRINT
+    dbg_printf("\r\nsleep...");
+#endif
+
+    _idle_until_alarm();
```

Comment:
If you want to do a light timed sleep, I think you can just call `mp_hal_delay_ms()`, which wakes up multiple times as necessary, subtracting the time most recently slept. It will also wake up on an interrupt, which I think would handle the pin case at the same time.